stock_picking_auto_print/models/stock_picking.py

```diff
@@ -0,0 +1,84 @@
+# Copyright (C) 2019 IBM Corp.
+# Copyright (C) 2019 Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+
+
+class StockPicking(models.Model):
+    _inherit = "stock.picking"
+
+    @api.multi
+    def _stock_picking_default_auto_print_report(self):
+        user_company_id = self.env.user.company_id.id
+        report_action_pool = self.env['ir.actions.report']
+        for picking in self.filtered(lambda p: p.sale_id):
+            default_report_id = False
+
+            # Find picking defaults reports
+            default_report_ids = report_action_pool.search(
+                [('model', '=', 'stock.picking'),
+                 ('is_default_report', '=', True)]).ids
+
+            # Check Partner country id
+            country_id = False
+            if picking.partner_id.country_id:
+                country_id = picking.partner_id.country_id.id
+
+            if default_report_ids and country_id:
+                # Filter report with Country and Company
+                report_ids = report_action_pool.search(
+                    [('country_id', '=', country_id),
+                     ('company_id', '=', user_company_id),
+                     ('id', 'in', default_report_ids)],
+                    limit=1)
+                if report_ids:
+                    default_report_id = report_ids.id
+
+            if not default_report_id and country_id:
+                # Filter report with Country
+                report_ids = report_action_pool.search(
+                    [('country_id', '=', country_id),
+                     ('id', 'in', default_report_ids)],
+                    limit=1)
+                if report_ids:
+                    default_report_id = report_ids.id
+
+            if not default_report_id:
+                # Filter report with Company
+                report_ids = report_action_pool.search(
+                    [('company_id', '=', user_company_id),
+                     ('id', 'in', default_report_ids)],
+                    limit=1)
+                if report_ids:
+                    default_report_id = report_ids.id
+
+            if not default_report_id:
+                default_report_id = self.env.ref(
+                    'stock.action_report_picking').with_context(
+                    landscape=True).report_action(picking).get('id')
+            action_report = report_action_pool.browse(
+                default_report_id)
```

Comment:
@bodedra Why not using search_read?

```
search_read(['company_id', 'country_id'], [('model', '=', 'stock.picking'), ('is_default_report', '=', True)])
```

and then use the results to determine which one to use.

Reply:
@max3903 I like the idea. But there is sequential order given to fetch default report. Here is sequence:

1. Country and Company
2. Country,
3. Company.

If not found, call "Picking Operations" report.

Is there way, we can identify our need using "search_read"?

Reply:
@bodedra Yes, instead of asking the DB and potentially running up to 4 SQL queries, use the list of dictionary returned by search_read:

```
[{'id': 1, 'company_id': 2, 'country_id': 3},
{'id': 4, 'company_id': 5, 'country_id': False},
{'id': 6, 'company_id': False, 'country_id': 7},
{'id': 8, 'company_id': False, 'country_id': False}]
```

Reply:
```
res = ...search_read(...)
for row in res:
    if row['company_id] == ... and row['country_id'] == ...:
        ...
    elseif row['country_id'] == ...:
        ...
    elseif row['company_id'] == ...:
        ..
    else:
        ...
browse
```

Reply:
@max3903 Make sense. :+1: